DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,927 @@
+value would require a copy of the `struct` which would be made by calling the copy constructor, thus leading
+to an infinite chain of calls).
+
+The type qualifiers may be applied to the parameter of the copy constructor, but also to the function itself
+in order to facilitate the ability to describe copies between objects of different mutability levels. The type
+qualifiers are optional.
+
+### Semantics
+
+This sections discusses all the aspects regarding the semantic analysis and interaction with other
+components of the copy constructor
+
+#### Copy constructor usage
+
+The copy constructor is implicitly used by the compiler whenever a struct variable is initialized
+as a copy of another variable of the same unqualified type:
+
+1. When a variable is explicitly initialized:
+
+```d
+struct A
+{
+    @implicit this(ref A another) {}
+}
+
+void main()
+{
+    A a;
+    A b = a; // copy constructor gets called
+    b = a;   // assignment, not initialization
+}
+```
+
+2. When a parameter is passed by value to a function:
+
+```d
+struct A
+{
+    @implicit this(ref A another) {}
+}
+
+void fun(A a) {}
+
+void main()
+{
+    A a;
+    fun(a);    // copy constructor gets called
+}
+```
+
+3. When a parameter is returned by value from a function:
+
+```d
+struct A
+{
+    @implicit this(ref A another) {}
+}
+
+A fun()
+{
+    A a;
+    return a;
+}
+
+void main()
+{
+    A a;
+    a = fun();      // copy constructor gets called
+}
+```
+
+The parameter of the copy constructor is passed by reference, so initializations are going
+to be lowered to copy constructor calls only if the source is an lvalue. Although this can be
+worked around by declaring temporary lvalues which can be forwarded to the copy constructor,
+binding rvalues to lvalues is beyond the scope of this DIP.
+
+Note that when a function returns a struct that defines a copy constructor, the copy constructor
+is called on the callee site, before actually returning:
+
+```d
+struct A
+{
+    @implicit this(ref A rhs) {}
+}
+
+A fun()
+{
+    A a;
+    return a;      // lowered to return tmp.copyCtor(a)
+    // return A();  // rvalue, no copyCtor called
+}
+
+void main()
+{
+    A b = fun();    // the return value of fun() is moved to the location of b
+}
+```
+
+#### Requirements
+
+1. The type of the parameter to the copy constructor needs to be identical to `typeof(this)`; an error is issued
+otherwise:
+
+```d
+struct A
+{
+    int a;
+    string b;
+}
+
+struct C
+{
+    A a;
+    string b;
+    @implicit this(ref C another)       // ok, typeof(this) == C
+    {
+        this.a = another.a;
+        this.b = another.b;
+    }
+
+    @implicit this(ref A another)       // error typeof(this) != C
+    {
+        this.a = another;
+        this.b = "hello";
+    }
+}
+
+void main()
+{
+    C c;
+    C d = c;   // ok
+    A a;
+    C e = a;   // error
+}
+```
+2. It is illegal to declare a copy constructor for a struct that has a postblit defined and vice versa:
```

Comment:
This requirement makes it impossible to transition to this DIP, I suggest a transitional `-dip10xx` flag that, if set, will prefer to call a copy constructor over a postblit (probably along with a `-vpostblit` to make sure that you're doing what you think you are, especially w.r.t qualifier combinations). It will also make it easier to keep compatibility with older compilers, which is very important.

Reply:
More that that, unless the `-dip10xx`flag defines a version (or one defines their own to ease transition) you can't have the two defined which makes it much more difficult to switch between  if, for some reason, your code breaks. It also means that you must version your code in order to compile with older compilers.

Reply:
> because a struct "has" a postblit even if one of its members does and "it" actually does not 

That doesn't help either.